File: MacOS/Type4Me/LLM/Providers/OpenAICompatibleLLMConfig.swift
import Foundation

// MARK: - Tag Protocol

/// Each OpenAI-compatible provider gets a zero-cost tag type
/// so the generic `OpenAICompatibleLLMConfig<Tag>` can map to the right `LLMProvider`.
protocol OpenAICompatibleLLMTag: Sendable {
    static var provider: LLMProvider { get }
}

// MARK: - Tags

enum DoubaoLLMTag:      OpenAICompatibleLLMTag { static let provider = LLMProvider.doubao }
enum MinimaxCNLLMTag:   OpenAICompatibleLLMTag { static let provider = LLMProvider.minimaxCN }
enum MinimaxIntlLLMTag: OpenAICompatibleLLMTag { static let provider = LLMProvider.minimaxIntl }
enum BailianLLMTag:     OpenAICompatibleLLMTag { static let provider = LLMProvider.bailian }
enum KimiLLMTag:        OpenAICompatibleLLMTag { static let provider = LLMProvider.kimi }
enum OpenRouterLLMTag:  OpenAICompatibleLLMTag { static let provider = LLMProvider.openrouter }
enum OpenAILLMTag:      OpenAICompatibleLLMTag { static let provider = LLMProvider.openai }
enum GeminiLLMTag:      OpenAICompatibleLLMTag { static let provider = LLMProvider.gemini }
enum DeepSeekLLMTag:    OpenAICompatibleLLMTag { static let provider = LLMProvider.deepseek }
enum ZhipuLLMTag:       OpenAICompatibleLLMTag { static let provider = LLMProvider.zhipu }

// MARK: - Generic Config

struct OpenAICompatibleLLMConfig<Tag: OpenAICompatibleLLMTag>: LLMProviderConfig, Sendable {

    static var provider: LLMProvider { Tag.provider }

    static var credentialFields: [CredentialField] {
        let p = Tag.provider
        return [
            CredentialField(
                key: "apiKey", label: "API Key",
                placeholder: "sk-...",
                isSecure: true, isOptional: false, defaultValue: ""
            ),
            CredentialField(
                key: "model", label: L("模型", "Model"),
                placeholder: L("模型名称或 endpoint ID", "Model name or endpoint ID"),
                isSecure: false, isOptional: false, defaultValue: ""
            ),
            CredentialField(
                key: "baseURL", label: "Base URL",
                placeholder: p.defaultBaseURL,
                isSecure: false, isOptional: true, defaultValue: p.defaultBaseURL
            ),
        ]
    }

    let apiKey: String
    let model: String
    let baseURL: String

    init?(credentials: [String: String]) {
        guard let key = credentials["apiKey"], !key.isEmpty,
              let model = credentials["model"], !model.isEmpty
        else { return nil }
        self.apiKey = key
        self.model = model
        let url = credentials["baseURL"]?.isEmpty == false
            ? credentials["baseURL"]!
            : Tag.provider.defaultBaseURL
        self.baseURL = url
    }

    func toCredentials() -> [String: String] {
        ["apiKey": apiKey, "model": model, "baseURL": baseURL]
    }

    func toLLMConfig() -> LLMConfig {
        LLMConfig(apiKey: apiKey, model: model, baseURL: baseURL)
    }
}

File: Windows/Type4Me/LLM/Providers/OpenAICompatibleLLMConfig.cs
using Type4Me.ASR;
using Type4Me.Localization;
using Type4Me.Models;

namespace Type4Me.LLM.Providers;

/// <summary>
/// Tag interface for OpenAI-compatible LLM providers.
/// </summary>
public interface IOpenAICompatibleLLMTag
{
    static abstract LLMProvider Provider { get; }
}

// ── Tag Types (one per provider) ───────────────────────

public struct DoubaoTag      : IOpenAICompatibleLLMTag { public static LLMProvider Provider => LLMProvider.Doubao; }
public struct MinimaxCNTag   : IOpenAICompatibleLLMTag { public static LLMProvider Provider => LLMProvider.MinimaxCN; }
public struct MinimaxIntlTag : IOpenAICompatibleLLMTag { public static LLMProvider Provider => LLMProvider.MinimaxIntl; }
public struct BailianTag     : IOpenAICompatibleLLMTag { public static LLMProvider Provider => LLMProvider.Bailian; }
public struct KimiTag        : IOpenAICompatibleLLMTag { public static LLMProvider Provider => LLMProvider.Kimi; }
public struct OpenRouterTag  : IOpenAICompatibleLLMTag { public static LLMProvider Provider => LLMProvider.OpenRouter; }
public struct OpenAITag      : IOpenAICompatibleLLMTag { public static LLMProvider Provider => LLMProvider.OpenAI; }
public struct GeminiTag      : IOpenAICompatibleLLMTag { public static LLMProvider Provider => LLMProvider.Gemini; }
public struct DeepSeekTag    : IOpenAICompatibleLLMTag { public static LLMProvider Provider => LLMProvider.DeepSeek; }
public struct ZhipuTag       : IOpenAICompatibleLLMTag { public static LLMProvider Provider => LLMProvider.Zhipu; }

/// <summary>
/// Generic config for all OpenAI-compatible LLM providers.
/// The TTag type parameter selects the provider.
/// </summary>
public sealed class OpenAICompatibleLLMConfig<TTag> where TTag : IOpenAICompatibleLLMTag
{
    public string ApiKey { get; }
    public string Model { get; }
    public string BaseURL { get; }

    public static LLMProvider Provider => TTag.Provider;

    public static CredentialField[] CredentialFields =>
    [
        new() { Key = "apiKey", Label = "API Key", Placeholder = "sk-...", IsSecure = true },
        new() { Key = "model", Label = Loc.L("模型", "Model"),
                Placeholder = Loc.L("模型名称或 endpoint ID", "Model name or endpoint ID") },
        new() { Key = "baseURL", Label = "Base URL",
                Placeholder = TTag.Provider.DefaultBaseURL(),
                IsOptional = true, DefaultValue = TTag.Provider.DefaultBaseURL() },
    ];

    private OpenAICompatibleLLMConfig(string apiKey, string model, string baseURL)
    {
        ApiKey = apiKey; Model = model; BaseURL = baseURL;
    }

    public static OpenAICompatibleLLMConfig<TTag>? TryCreate(Dictionary<string, string> credentials)
    {
        if (!credentials.TryGetValue("apiKey", out var key) || string.IsNullOrEmpty(key)) return null;
        if (!credentials.TryGetValue("model", out var model) || string.IsNullOrEmpty(model)) return null;
        var baseURL = credentials.GetValueOrDefault("baseURL");
        if (string.IsNullOrEmpty(baseURL)) baseURL = TTag.Provider.DefaultBaseURL();
        return new OpenAICompatibleLLMConfig<TTag>(key, model, baseURL);
    }

    public Dictionary<string, string> ToCredentials() => new()
    {
        ["apiKey"] = ApiKey, ["model"] = Model, ["baseURL"] = BaseURL,
    };

    public LLMConfig ToLLMConfig() => new(ApiKey, Model, BaseURL);
}
